ROCpd schema fetching from rocprofiler-sdk

mradosav-amd opened the pull request:
## Motivation
Fetch ROCpd schema content from rocprofiler-sdk-rocpd, if `ROCPROFSYS_USE_ROCM` is set to `ON`.
In other case, use schema which is embedded in source code.

## Technical Details
- Integrate rocprofiler-systems with rocprofiler-sdk-rocpd to fetch schema
- If rocprofiler-sdk-rocpd is not availabe, use embedded schema files. With this we provide rocpd format support even if ROCm is not available
- Include detection in CMake if rocprofiler-sdk-rocpd package is available (and valid), and build database class upon that
- Update embedded schema that is used as a fallback.

## Test Plan
Test with:
- ROCm 6.4.1 (no rocprofiler-sdk-rocpd available)
- ROCm 7.0.1 (rocprofiler-sdk-rocpd do not contain sql.h file, still not valid) 
- ROCm 7.0.2 (rocprofiler-sdk-rocpd is valid)

## Test Result

Tests pass.

## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

mradosav-amd committed:
Add schema fetching from rocprofiler-sdk

mradosav-amd committed:
Schema fetching

mradosav-amd committed:
Cmake cleanup

mradosav-amd committed:
Add changelog

kcossett-amd commented on projects/rocprofiler-systems/source/lib/core/rocpd/data_storage/database.cpp:
```diff
@@ -32,6 +32,29 @@
+    ROCPD_SQL_SCHEMA_ROCPD_MARKER_VIEWS,
+    ROCPD_SQL_SCHEMA_ROCPD_SUMMARY_VIEWS
+};
+};
```
Semicolon after closing bracket for anonymous namespace is not needed. It creates a warning causing certain workflows to fail.

mradosav-amd replied:
Done

kcossett-amd commented on projects/rocprofiler-systems/source/lib/core/rocpd/data_storage/database.cpp:
```diff
@@ -32,6 +32,29 @@
+#if defined(ROCPROFSYS_USE_ROCPD_LIBRARY) && ROCPROFSYS_USE_ROCPD_LIBRARY > 0
+#    include <rocprofiler-sdk-rocpd/rocpd.h>
+#else
+#    include "core/rocpd/data_storage/schema/data_views.hpp"
```
When building with `ROCPROFSYS_USE_ROCM=OFF`, it cannot find the file. See the failed `ubuntu-noble` workflow (at line 7771)

```
fatal error: core/rocpd/data_storage/schema/data_views.hpp: No such file or directory
   38 | #    include "core/rocpd/data_storage/schema/data_views.hpp"
```

mradosav-amd committed:
CMake refactor

adjordje-amd commented on projects/rocprofiler-systems/CHANGELOG.md:
```diff
@@ -10,6 +10,7 @@ Full documentation for ROCm Systems Profiler is available at [https://rocm.docs.
 
 - Added a `ROCPROFSYS_PERFETTO_FLUSH_PERIOD_MS` configuration setting to set the flush period for Perfetto traces. The default value is 10000 ms (10 seconds).
 - Added tracing of VAAPI, MPI and host events inside `rocpd` output database.
+- Added fetching of `ROCpd` schema from rocprofiler-sdk-rocpd
```
```suggestion
- Added fetching of `ROCpd` schema from rocprofiler-sdk-rocpd
- Added fetching of the `ROCpd` schema from rocprofiler-sdk-rocpd
```

adjordje-amd commented on projects/rocprofiler-systems/cmake/Packages.cmake:
```diff
@@ -232,6 +232,97 @@ if(ROCPROFSYS_USE_ROCM)
+find_package(rocprofiler-sdk-rocpd ${rocprofiler_systems_FIND_QUIETLY})
+
+if(rocprofiler-sdk-rocpd_FOUND)
+    set(ROCPROFSYS_ROCPD_HAS_SQL_H FALSE)
```
I think you can remove the `ROCPROFSYS_ROCPD_HAS_SQL_H` flag because support for schema fetching from the rocpd library has already been merged with a fix for the missing header, so we should be on safe ground

adjordje-amd commented on projects/rocprofiler-systems/source/lib/core/rocpd/data_storage/database.cpp:
```diff
@@ -44,7 +66,65 @@ create_directory_for_database_file(const std::string& db_file)
+get_schema_query(rocpd_sql_schema_kind_t schema_kind, const std::string& upid)
+{
+#if defined(ROCPROFSYS_USE_ROCPD_LIBRARY) && ROCPROFSYS_USE_ROCPD_LIBRARY > 0
+    rocpd_sql_schema_jinja_variables_t info{ 2 * upid.size(), upid.c_str(),
```
Replace the expression `2 * upid.size()` with a clearly named variable

adjordje-amd reviewed (commented):
I don’t like the sql.h existence check, but we must do it to support ROCm 7.0. I also dislike the ifdefs, but they’re necessary because of ROCPROFSYS_USE_ROCM and the need for the schema even when the ROCm stack is unavailable.

LGTM

mradosav-amd committed:
Applied suggestions from code review

dgaliffiAMD commented on projects/rocprofiler-systems/CHANGELOG.md:
```diff
@@ -10,6 +10,7 @@ Full documentation for ROCm Systems Profiler is available at [https://rocm.docs.
 
 - Added a `ROCPROFSYS_PERFETTO_FLUSH_PERIOD_MS` configuration setting to set the flush period for Perfetto traces. The default value is 10000 ms (10 seconds).
 - Added tracing of VAAPI, MPI and host events inside `rocpd` output database.
+- Added fetching of the `ROCpd` schema from rocprofiler-sdk-rocpd
```
```suggestion
- Added fetching of the `rocpd` schema from `rocprofiler-sdk-rocpd`.
```

Matching casing of "rocpd" used elsewhere in the document.

mradosav-amd replied:
Done

dgaliffiAMD commented on projects/rocprofiler-systems/cmake/Packages.cmake:
```diff
@@ -232,6 +232,97 @@ if(ROCPROFSYS_USE_ROCM)
+        ROCPD_INCLUDE_DIRS
+        rocprofiler-sdk-rocpd::rocprofiler-sdk-rocpd
+        INTERFACE_INCLUDE_DIRECTORIES
+    )
```
I'm building with rocm-7.1, but ROCPD_INCLUDE_DIRS is not found. It's always going down the "fallback" path. 

```
-- Found rocprofiler-sdk-rocpd: /opt/rocm-7.1.0 (found version "1.0.0")  
-- ROCPD_INCLUDE_DIRS: ROCPD_INCLUDE_DIR-NOTFOUND
-- rocprofiler-sdk-rocpd_INCLUDE_DIR: /opt/rocm-7.1.0/include
-- rocprofiler-sdk-rocpd_LIBRARIES: rocprofiler-sdk-rocpd::rocprofiler-sdk-rocpd-shared-library
-- rocprofiler-sdk-rocpd_ROOT_DIR: /opt/rocm-7.1.0
-- rocprofiler-sdk-rocpd not found - using local schema files
```

mradosav-amd replied:
Fixed.

dgaliffiAMD commented:
It looks like the views in the schema have changed since our local copy, which is causing some of our rocpd validation tests to fail. There is no longer a "kernel_summary" view.

mradosav-amd committed:
Applied suggestions from code review

dgaliffiAMD commented on projects/rocprofiler-systems/source/lib/core/rocpd/data_storage/database.cpp:
```diff
@@ -44,7 +66,65 @@ create_directory_for_database_file(const std::string& db_file)
+    rocpd_sql_schema_jinja_variables_t info{ jinja_size, upid.c_str(), upid.c_str() };
+
+    std::string query;
+    rocpd_sql_load_schema(ROCPD_SQL_ENGINE_SQLITE3, schema_kind, ROCPD_SQL_OPTIONS_NONE,
```
It would probably be a good idea to check the return code here.

mradosav-amd replied:
Done.

dgaliffiAMD commented on projects/rocprofiler-systems/source/lib/core/rocpd/data_storage/database.cpp:
```diff
@@ -44,7 +66,65 @@ create_directory_for_database_file(const std::string& db_file)
+               const rocpd_sql_schema_jinja_variables_t*, const char*,
+               const char* schema_content, void* user_data)
+{
+    auto* query = static_cast<std::string*>(user_data);
```
Right now the callback is pretty simple; the only user of this function is right below. For future, do you think you should add a null-check before dereferencing and assigning `*query`.

mradosav-amd replied:
Guards added.

dgaliffiAMD commented on projects/rocprofiler-systems/source/lib/core/rocpd/data_storage/database.cpp:
```diff
@@ -44,7 +66,65 @@ create_directory_for_database_file(const std::string& db_file)
+          rocprofsys::rocpd::data_storage::schema::MARKER_VIEWS_SQL },
+        { ROCPD_SQL_SCHEMA_ROCPD_SUMMARY_VIEWS,
+          rocprofsys::rocpd::data_storage::schema::SUMMARY_VIEWS_SQL }
+    };
```
Just a suggestion: It may be overkill to use a map, when a simple switch / case may do and provide less overhead?

mradosav-amd replied:
Sure, it would be faster with switch case.

dgaliffiAMD committed:
Update rocpd validation tables.

Using the `top_kernel` view, rather than `kernel_summary`, which has been removed

mradosav-amd committed:
Remove std::map, fix cmake

mradosav-amd committed:
Cmake cleanup

mradosav-amd committed:
Add guards for pointers

dgaliffiAMD commented on projects/rocprofiler-systems/cmake/Packages.cmake:
```diff
@@ -232,6 +232,110 @@ if(ROCPROFSYS_USE_ROCM)
+if(rocprofiler-sdk-rocpd_FOUND)
+    set(ROCPROFSYS_ROCPD_HAS_SQL_H FALSE)
+
+    if(rocprofiler-sdk-rocpd_DIR)
```
Could you use `INCLUDE_DIR` or `ROOT_DIR` here instead? This will save you all of the string manipulation through `get_filename_component`.

```suggestion
    if(rocprofiler-sdk-rocpd_INCLUDE_DIR)
```

For reference, 
```
-- rocprofiler-sdk-rocpd_FOUND: 1
-- rocprofiler-sdk-rocpd_DIR: /opt/rocm-7.1.0/lib/cmake/rocprofiler-sdk-rocpd
-- rocprofiler-sdk-rocpd_VERSION: 1.0.0
-- rocprofiler-sdk-rocpd_INCLUDE_DIR: /opt/rocm-7.1.0/include
-- rocprofiler-sdk-rocpd_LIBRARIES: rocprofiler-sdk-rocpd::rocprofiler-sdk-rocpd-shared-library
-- rocprofiler-sdk-rocpd_ROOT_DIR: /opt/rocm-7.1.0
-- rocprofiler-sdk-rocpd_LIB_DIR: /opt/rocm-7.1.0/lib
```

dgaliffiAMD commented on projects/rocprofiler-systems/cmake/Packages.cmake:
```diff
@@ -232,6 +232,110 @@ if(ROCPROFSYS_USE_ROCM)
+            _SUFFIX
+            "include"
+            "include/rocprofiler-sdk-rocpd"
+            "include/rocprofiler-sdk"
```
I'm not sure that you have to iterate through all these possibilities. If you do, then I would put `rocprofiler-sdk-rocpd` at the top because it is the most likely location and that will reduce iterations.

```suggestion
            "include/rocprofiler-sdk-rocpd"
            "include/rocprofiler-sdk"
            "include"
```

mradosav-amd committed:
Applied suggestions from code review
